### ISSUE TITLE
Change from Function et al. interfaces + magic to input functions struct

### ISSUE THREAD

**Issue by vladimir-ch** [gonum/optimize#102]
Less magic and more composibility by switching to something like what @btracey proposes:

```go
type Problem struct {
    Func func([]float64) float64
    Grad func([]float64, []float64)
    Hess func([]float64, *mat64.SymDense)

    Constraints func(x, equality, inequality []float64)
    ConstraintGrad func(x []float64, eq, ineq *mat64.Dense) // maybe Matrix
    LinearEq struct{A mat64.Dense, b []float64}
    LinearIneq struct{A mat64.Dense, b []float64}
    Bounds []struct{Min, Max float64}
}
```

**Comment by btracey** [comment]
This should have
````
type Bound struct{Min float64; Max float64}
````
and 
````
type LinearConstraint struct{
A *mat64.Dense
b []float64
}
````
instead of the anonymous types.

**Comment by btracey** [comment]
This currently has a problem in that there is no way to communicate how many non-linear constraints there are. We could have a NumConstraints(), but that would be unfortunate. Maybe just mandate that Constraints has to allow nil arguments the first time.

**Comment by hypoactiv** [comment]
I took an initial stab at this on my fork at https://github.com/hypoactiv/optimize today. Comments and criticism are welcome. A few notes on my changes:

- I didn't yet do anything to update the testing system to the new style, so 'go test' won't run.
- I removed the FunctionInfo struct that was originally being passed around, since it seemed redundant after introducing the Problem struct. Instead of using funcInfo.IsGradient, etc, I check if p.Grad != nil.
- The Problem struct currently doesn't have any fields for constraints.
